File: src/main/java/org/distril/beengine/network/data/transaction/ItemUseTransaction.java
package org.distril.beengine.network.data.transaction;

import com.nukkitx.math.vector.Vector3f;
import com.nukkitx.math.vector.Vector3i;
import com.nukkitx.protocol.bedrock.packet.InventoryTransactionPacket;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import org.distril.beengine.material.item.Item;
import org.distril.beengine.util.Direction;
import org.distril.beengine.util.ItemUtil;

@Getter
@RequiredArgsConstructor
public class ItemUseTransaction {

	private final ItemUseType type;

	private final Vector3i blockPosition;
	private final Direction blockFace;

	private final int hotbarSlot;
	private final Item itemInHand;

	private final Vector3f playerPosition, clickPosition;

	private final int blockRuntimeId;

	public static ItemUseTransaction read(InventoryTransactionPacket packet) {
		var type = ItemUseType.fromTypeId(packet.getActionType());

		var blockPosition = packet.getBlockPosition();
		var blockFace = Direction.fromIndex(packet.getBlockFace());

		var hotbarSlot = packet.getHotbarSlot();
		var itemInHand = ItemUtil.fromNetwork(packet.getItemInHand());

		var playerPosition = packet.getPlayerPosition();
		var clickPosition = packet.getClickPosition();

		var blockRuntimeId = packet.getBlockRuntimeId();
		return new ItemUseTransaction(type, blockPosition, blockFace, hotbarSlot, itemInHand, playerPosition,
				clickPosition, blockRuntimeId);
	}

	public enum ItemUseType {

		CLICK_BLOCK,
		CLICK_AIR,
		BREAK_BLOCK;

		public static ItemUseType fromTypeId(int typeId) {
			return ItemUseType.values()[typeId];
		}
	}
}

File: src/main/kotlin/org/distril/beengine/network/data/transaction/ItemUseTransaction.kt
package org.distril.beengine.network.data.transaction

import com.nukkitx.math.vector.Vector3f
import com.nukkitx.math.vector.Vector3i
import com.nukkitx.protocol.bedrock.packet.InventoryTransactionPacket
import org.distril.beengine.material.item.Item
import org.distril.beengine.util.Direction
import org.distril.beengine.util.ItemUtils

class ItemUseTransaction private constructor(
	val type: Type,
	val blockPosition: Vector3i,
	val blockFace: Direction,
	val hotbarSlot: Int,
	val itemInHand: Item,
	val playerPosition: Vector3f,
	val clickPosition: Vector3f,
	val blockRuntimeId: Int
) {

	enum class Type {

		CLICK_BLOCK,
		CLICK_AIR,
		BREAK_BLOCK;

		companion object {

			fun fromTypeId(typeId: Int) = Type.values()[typeId]
		}
	}

	companion object {

		fun read(packet: InventoryTransactionPacket): ItemUseTransaction {
			val type = Type.fromTypeId(packet.actionType)
			val blockPosition = packet.blockPosition
			val blockFace = Direction.fromIndex(packet.blockFace)!!
			val hotbarSlot = packet.hotbarSlot
			val itemInHand = ItemUtils.fromNetwork(packet.itemInHand)
			val playerPosition = packet.playerPosition
			val clickPosition = packet.clickPosition
			val blockRuntimeId = packet.blockRuntimeId
			return ItemUseTransaction(
				type, blockPosition, blockFace, hotbarSlot, itemInHand, playerPosition,
				clickPosition, blockRuntimeId
			)
		}
	}
}
